core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java

```diff
@@ -0,0 +1,211 @@
+
+        List<String> command = new ArrayList<>(4);
+        command.add(javaBin);
+        command.add("-XX:DumpLoadedClassList=" + classesLstPath.toAbsolutePath().toString());
+        command.add("-jar");
+        command.add(jarResult.getPath().toAbsolutePath().toString());
+
+        if (log.isDebugEnabled()) {
+            log.debugf("Launching command: '%s' to create class list", String.join(" ", command));
+        }
+        try {
+            Process process = new ProcessBuilder(command)
+                    .redirectErrorStream(true)
+                    .directory(outputTarget.getOutputDirectory().toFile())
+                    .start();
+            waitForApplicationToStart(process);
+        } catch (IOException e) {
+            log.debug("Failed to launch process used to create 'classes.lst'.", e);
+            return null;
+        }
+
+        if (!classesLstPath.toFile().exists()) { // shouldn't happen, but let's avoid any surprises
+            return null;
+        }
+
+        return classesLstPath;
+    }
+
+    /**
+     * Waits for the application terminate by looking for the proper logs, or just giving up after some time
+     */
+    private void waitForApplicationToStart(Process process) throws IOException {
+        if (!process.isAlive()) {
+            log.debug("The process that was supposed to create 'classes.lst' exited.");
+            return;
+        }
+        // don't wait for more than 10 seconds
+        // TODO: make it configurable?
+        final int maxWaits = 200;
+        int waits = 0;
+        BufferedReader in = new BufferedReader(new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
+        while (true) {
+            if (!process.isAlive()) {
+                log.debug("The process that was supposed to create classes.lst exited");
+                break;
+            }
+            String line = in.readLine();
+            if (line != null) {
+                if (line.contains(Timing.INSTALLED_FEATURES) || line.contains(MainClassBuildStep.FAILED_TO_START_MESSAGE)) {
+                    process.destroy();
+                    break;
+                }
+            }
+            try {
+                Thread.sleep(50);
+            } catch (InterruptedException ignored) {
+            }
+
+            // we don't want to wait forever
+            if (++waits >= maxWaits) {
+                break;
+            }
+        }
+        try {
+            in.close();
+        } catch (IOException ignored) {
+        }
+    }
+
+    /**
+     * @return The path of the created app-cds.jsa file or null of the file was not created
+     */
+    private Path createAppCDS(JarBuildItem jarResult, String javaBin,
+            Path classesLstPath, boolean isFastFar) {
+
+        Path workingDirectory = isFastFar ? jarResult.getPath().getParent().getParent() : jarResult.getPath().getParent();
+        Path appCDSPath = workingDirectory.resolve("app-cds.jsa");
```

Comment:
For fast-jar this should go into the actual fast-jar package.

Reply:
The way it's done, it ends up in the directory that `quarkus-run.jar` exists. Is that what you meant, or something else?

Reply:
That is good, I missread this bit.

Reply:
That is good, I missread this bit.